File: clojure/bob/src/bob.clj
(ns bob (:require [clojure.string :as str]))

(defn- question? [s] (str/ends-with? s "?"))

(defn- shouting? [s] (and (re-find #"[a-zA-Z]" s) (= s (str/upper-case s))))

(defn response-for [s]
  (let [cleaned (str/trim s)]
    (cond
      (empty? cleaned) "Fine. Be that way!"
      (and (shouting? cleaned) (question? cleaned)) "Calm down, I know what I'm doing!"
      (shouting? cleaned) "Whoa, chill out!"
      (question? cleaned) "Sure."
      :else "Whatever.")))

File: elixir/bob/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    trimmed = String.trim(input)
    cond do
      shouting?(trimmed) and question?(trimmed) -> "Calm down, I know what I'm doing!"
      silence?(trimmed) -> "Fine. Be that way!"
      question?(trimmed) -> "Sure."
      shouting?(trimmed) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp shouting?(input), do: String.upcase(input) == input and String.downcase(input) != input

  defp question?(input), do: String.ends_with?(input, "?")

  defp silence?(input), do: String.trim(input) == ""
end
